File: 18-07-2024_Number_of_Good_Leaf_Nodes_Pairs.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {
    int ans = 0;

    public List<Integer> dfs(TreeNode root, int d) {
        if (root == null) return new ArrayList<>();

        if (root.left == null && root.right == null) {
            List<Integer> leafList = new ArrayList<>();
            leafList.add(1);
            return leafList;
        }

        List<Integer> left = dfs(root.left, d);
        List<Integer> right = dfs(root.right, d);

        for (int a : left) {
            for (int b : right) {
                if (a + b <= d) {
                    ans++;
                }
            }
        }

        List<Integer> v = new ArrayList<>();
        for (int lv : left) {
            if (lv + 1 <= d) {
                v.add(lv + 1);
            }
        }
        for (int rv : right) {
            if (rv + 1 <= d) {
                v.add(rv + 1);
            }
        }

        return v;
    }

    public int countPairs(TreeNode root, int distance) {
        dfs(root, distance);
        return ans;
    }
}

File: 18-07-2024_Number_of_Good_Leaf_Nodes_Pairs.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    ans = 0
    def dfs(self, root, d):
        if not root:
            return []

        if not root.left and not root.right:
            return [1]

        left = self.dfs(root.left, d)
        right = self.dfs(root.right, d)

        for a in left:
            for b in right:
                if a + b <= d:
                    self.ans += 1

        v = []
        for lv in left:
            if lv + 1 <= d:
                v.append(lv + 1)
        for rv in right:
            if rv + 1 <= d:
                v.append(rv + 1)

        return v

    def countPairs(self, root, distance):
        self.dfs(root, distance)
        return self.ans
